lib/membrane_h264_plugin/parser.ex

```diff
@@ -129,51 +180,77 @@ defmodule Membrane.H264.Parser do
       au_counter: 0,
       output_alignment: opts.output_alignment,
       frame_prefix: <<>>,
-      parameter_sets_present?: byte_size(opts.sps) > 0 or byte_size(opts.pps) > 0,
       skip_until_keyframe?: opts.skip_until_keyframe?,
       repeat_parameter_sets?: opts.repeat_parameter_sets,
-      cached_sps: %{},
-      cached_pps: %{}
+      cached_spss: %{},
+      cached_ppss: %{},
+      initial_spss: initial_parameters_to_list(opts.sps),
+      initial_ppss: initial_parameters_to_list(opts.pps),
+      input_stream_structure: nil,
+      output_stream_structure: output_stream_structure
     }
 
     {[], state}
   end
 
   @impl true
-  def handle_stream_format(:input, stream_format, _ctx, state) do
-    state =
+  def handle_stream_format(:input, stream_format, ctx, state) do
+    {mode, input_raw_stream_structure} =
       case stream_format do
         %RemoteStream{type: :bytestream} ->
-          %{state | mode: :bytestream}
+          {:bytestream, :annexb}
+
+        %H264{alignment: alignment, stream_structure: stream_structure} ->
+          {alignment, stream_structure}
+      end
+
+    mode =
+      case mode do
+        :au -> :au_aligned
+        :nalu -> :nalu_aligned
+        :bytestream -> :bytestream
+      end
 
-        %H264.RemoteStream{alignment: alignment, decoder_configuration_record: dcr} ->
-          mode =
-            case alignment do
-              :nalu -> :nalu_aligned
-              :au -> :au_aligned
-            end
+    first_received_stream_format? = is_nil(ctx.pads.input.stream_format)
 
-          frame_prefix =
-            case dcr do
-              nil ->
-                <<>>
+    state =
+      if first_received_stream_format? do
+        {input_stream_structure, output_stream_structure} =
+          get_stream_structures(input_raw_stream_structure, state.output_stream_structure)
 
-              _dcr when state.parameter_sets_present? ->
-                raise "Parameter sets were already provided as the options to the parser and parameter sets from the decoder configuration record could overwrite them."
+        %{
+          state
+          | mode: mode,
+            nalu_splitter: NALuSplitter.new(input_stream_structure),
+            nalu_parser: NALuParser.new(input_stream_structure, output_stream_structure),
+            input_stream_structure: input_stream_structure,
+            output_stream_structure: output_stream_structure
+        }
+      else
+        if not stream_structures_compatible?(
```

Comment:
I would call this function more like `is_input_stream_structure_change_allowed?` as this is what it in fact does. When I've first taken a look at `stream_structures_compatible?` I though that we are restricting the input <-> output conversion.